decidim-ldap/lib/decidim/ldap/extensions/authentication_strategy.rb

```diff
@@ -0,0 +1,65 @@
+require 'net/ldap'
+
+# rubocop:disable Metrics/BlockLength
+Warden::Strategies.add(:ldap_authenticatable) do
+  def authenticate!
+    return unless params[:user]
+
+    ldap = Net::LDAP.new
+    ldap.host = ldap_configuration.host
+    ldap.port = ldap_configuration.port
+    ldap.base = ldap_configuration.authentication_query_for_username(username)
+    ldap.auth(ldap.base, password)
+
+    return fail unless ldap.bind
+
+    user = find_or_create_user(ldap)
+    success!(user)
+  end
+
+  private
+
+  def ldap_configuration
+    @organization ||=
+      Decidim::Organization.find(params[:organization_id]).ldap_configuration
+  end
+
+  def username
+    params[:user][:name]
+  end
+
+  def password
+    params[:user][:password]
+  end
+
+  def find_or_create_user(ldap)
+    user = find_user
+    return user if user.present?
+
+    create_user(ldap)
+  end
+
+  def find_user
+    Decidim::User.find_by(name: username.downcase,
+                          organization: ldap_configuration.organization)
+  end
+
+  def create_user(ldap)
+    user = Decidim::User.new
+    user.email = ldap_field_value(ldap, ldap_configuration.email_field)
+    user.nickname = ldap_field_value(ldap, ldap_configuration.name_field)
+    user.password = Devise.friendly_token.first(8)
+    user.organization = ldap_configuration.organization
+    user.name = ldap_field_value(ldap, ldap_configuration.username_field)
+    user.tos_agreement = true
+    user.skip_confirmation!
+    user.locale = I18n.locale
+    user.save
+
+    user
+  end
+
+  def ldap_field_value(ldap, key)
+    ldap.search[0][key][0]
```

Comment:
If the user is an Admin, search will return the entire LDAP directory. For that reason, this is the place where we need to use the DN and the authentication query:

`ldap.search(base: dn, filter: authentication_query)`

2 more comments:

- The result for the search query can be nil. We need to control it.
- Everytime we call search, we are making a query against the LDAP. We need to cache the search call. And then use it to retrieve the values